REGRESSION(288001@main): [GLIB] Build fix when BUBBLEWRAP_SANDBOX is disabled

dpino opened the pull request:
#### a7d65dca86e3d82215569d65ceaaa4a71ae8e403
<pre>
REGRESSION(288001@main): [GLIB] Build fix when BUBBLEWRAP_SANDBOX is disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=280237">https://bugs.webkit.org/show_bug.cgi?id=280237</a>

Reviewed by Michael Catanzaro.

Also fix pre-existing file descriptor leak.

* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: <a href="https://commits.webkit.org/288083@main">https://commits.webkit.org/288083@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6dcbdfa0aeb6d881a0e7e3422326a4c52504780e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81862 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86416 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32858 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1421 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9210 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63847 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21576 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84932 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1046 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74495 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/44133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/945 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31311 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29279 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87846 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9102 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72200 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9287 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70316 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71429 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15524 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14443 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12679 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9053 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14585 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8894 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12417 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10702 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -114,12 +114,12 @@ void ProcessLauncher::launchProcess()
         WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
 
-        m_processID = ProcessProviderLibWPE::singleton().launchProcess(m_launchOptions, argv, WTFMove(webkitSocketPair.client));
+        m_processID = ProcessProviderLibWPE::singleton().launchProcess(m_launchOptions, argv, webkitSocketPair.client.release());
```
Hm, I see the original code doesn't use `webkitSocketPair.client` past this point, so clearly it was expecting to transfer ownership, and here you preserve the previous behavior. But `ProcessProviderLibWPE::launchProcess` (which is hidden, it's actually in ProcessProviderGLib.cpp) does not expect to receive ownership, so we actually leak the file descriptor. That's a pretty serious preexisting bug. I also checked the PlayStation interface and it also does not receive ownership, but it closes the client fd so doesn't have this bug.

 Might as well fix it now:

```suggestion
        m_processID = ProcessProviderLibWPE::singleton().launchProcess(m_launchOptions, argv, webkitSocketPair.client.value());
```

This is important and arguably we shouldn't sneak it into a build fix commit, so you could land it separately if you prefer, or edit the commit message to mention that you're fixing more than just the build regression, whichever way seems nicer to you.

mcatanzaro replied:
>  But ProcessProviderLibWPE::launchProcess (which is hidden, it's actually in ProcessProviderGLib.cpp) does not expect to receive ownership, so we actually leak the file descriptor.

This is somewhat of an understatement; it just completely ignores the client file descriptor. I do not understand how the launched process is able to function without its IPC connection to the parent process. That's _really_ weird. I wonder what I'm missing?

dpino replied:
I applied the suggested change and updated commit message.

mcatanzaro reviewed (approved):
Sorry for the breakage.

dpino committed:
REGRESSION(288001@main): [GLIB] Build fix when BUBBLEWRAP_SANDBOX is disabled
https://bugs.webkit.org/show_bug.cgi?id=280237

Reviewed by Michael Catanzaro.

Also fix pre-existing file descriptor leak.

* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: https://commits.webkit.org/288083@main
